airflow/utils/state.py

```diff
@@ -16,70 +16,110 @@
-    SCHEDULED = "scheduled"
+    # Set by the scheduler
+    # None - Task is created but should not run yet
+    REMOVED = "removed"  # Task vanished from DAG before it ran
+    SCHEDULED = "scheduled"  # Task should run and will be handed to executor soon
+
+    # Set by the task instance itself
+    QUEUED = "queued"  # Executor has enqueued the task
+    RUNNING = "running"  # Task is executing
+    SUCCESS = "success"  # Task completed
+    SHUTDOWN = "shutdown"  # External request to shut down
+    FAILED = "failed"  # Task errored out
+    UP_FOR_RETRY = "up_for_retry"  # Task failed but has retries left
+    UP_FOR_RESCHEDULE = "up_for_reschedule"  # A waiting `reschedule` sensor
+    UPSTREAM_FAILED = "upstream_failed"  # One or more upstream deps failed
+    SKIPPED = "skipped"  # Skipped by branching or some other mechanism
+    SENSING = "sensing"  # Smart sensor offloaded to the sensor DAG
+
+
+class DagState(str, Enum):
+    """
+    Enum that represents all possible states that a DagRun can be in.
 
-    # set by the executor (t.b.d.)
-    # LAUNCHED = "launched"
+    These are "shared" with TaskState in some parts of the code, so make
+    sure they don't drift.
+    """
 
-    # set by a task
-    QUEUED = "queued"
     RUNNING = "running"
     SUCCESS = "success"
-    SHUTDOWN = "shutdown"  # External request to shut down
     FAILED = "failed"
-    UP_FOR_RETRY = "up_for_retry"
-    UP_FOR_RESCHEDULE = "up_for_reschedule"
-    UPSTREAM_FAILED = "upstream_failed"
-    SKIPPED = "skipped"
-    SENSING = "sensing"
-
-    task_states = (
-        SUCCESS,
-        RUNNING,
-        FAILED,
-        UPSTREAM_FAILED,
-        SKIPPED,
-        UP_FOR_RETRY,
-        UP_FOR_RESCHEDULE,
-        QUEUED,
-        NONE,
-        SCHEDULED,
-        SENSING,
-        REMOVED,
+
+
+class State:
+    """
+    Static class with task instance states constants and color method to
+    avoid hardcoding.
+    """
+
+    # Backwards-compat constants for code that does not yet use the enum
+    # These first three are shared by DagState and TaskState
+    SUCCESS = TaskState.SUCCESS
+    RUNNING = TaskState.RUNNING
+    FAILED = TaskState.FAILED
+
+    # These are TaskState only
+    NONE = None
+    REMOVED = TaskState.REMOVED
+    SCHEDULED = TaskState.SCHEDULED
+    QUEUED = TaskState.QUEUED
+    SHUTDOWN = TaskState.SHUTDOWN
+    UP_FOR_RETRY = TaskState.UP_FOR_RETRY
+    UP_FOR_RESCHEDULE = TaskState.UP_FOR_RESCHEDULE
+    UPSTREAM_FAILED = TaskState.UPSTREAM_FAILED
+    SKIPPED = TaskState.SKIPPED
+    SENSING = TaskState.SENSING
+
+    task_states: Tuple[Optional[TaskState], ...] = (
+        None,
+        TaskState.SUCCESS,
+        TaskState.RUNNING,
+        TaskState.FAILED,
+        TaskState.UPSTREAM_FAILED,
+        TaskState.SKIPPED,
+        TaskState.UP_FOR_RETRY,
+        TaskState.UP_FOR_RESCHEDULE,
+        TaskState.QUEUED,
+        TaskState.SCHEDULED,
+        TaskState.SENSING,
+        TaskState.REMOVED,
```

Comment:
```suggestion
    task_states: Tuple[Optional[TaskState], ...] = (None,) + tuple(TaskState)
```

That way it can't ever get out of sync https://docs.python.org/3/library/enum.html#iteration